File: packages/w2v/reference/src/training/cbow.c
#include "atomic_float.h"
#include "training.h"

#include <string.h>

static void accumulate_input(
    const Model *model,
    size_t token,
    real *hidden)
{
    size_t dimension = model->embedding_dimension;
    const _Atomic uint32_t *input_row =
        model->input_embeddings + token * dimension;

    for (size_t coordinate = 0; coordinate < dimension; coordinate++)
    {
        hidden[coordinate] += atomic_float_load(&input_row[coordinate]);
    }
}

static void update_input(
    Model *model,
    size_t token,
    const real *gradient)
{
    size_t dimension = model->embedding_dimension;
    _Atomic uint32_t *input_row =
        model->input_embeddings + token * dimension;

    for (size_t coordinate = 0; coordinate < dimension; coordinate++)
    {
        atomic_float_add(&input_row[coordinate], gradient[coordinate]);
    }
}

void cbow_train(ModelStep *step)
{
    Model *model = step->trainer->model;
    Worker *worker = step->worker;
    size_t dimension = model->embedding_dimension;
    size_t radius = context_radius(step);
    size_t context_count = 0;

    memset(worker->hidden, 0, dimension * sizeof(*worker->hidden));
    memset(
        worker->hidden_gradient,
        0,
        dimension * sizeof(*worker->hidden_gradient));

    for (size_t offset = 0; offset <= radius * 2; offset++)
    {
        size_t position;
        if (context_position(worker, radius, offset, &position))
        {
            size_t token = worker->sentence[position];

            accumulate_input(model, token, worker->hidden);
            context_count++;
        }
    }
    if (context_count == 0)
    {
        return;
    }

    for (size_t coordinate = 0; coordinate < dimension; coordinate++)
    {
        worker->hidden[coordinate] /= (real)context_count;
    }
    objective_train(step, worker->hidden, worker->hidden_gradient);

    for (size_t offset = 0; offset <= radius * 2; offset++)
    {
        size_t position;
        if (context_position(worker, radius, offset, &position))
        {
            size_t token = worker->sentence[position];

            update_input(model, token, worker->hidden_gradient);
        }
    }
}

File: packages/w2v/src/training/cbow.rs
use super::{ModelStep, ObjectiveLoss, context_position, context_radius_with_policy, objective};
use crate::{
    atomic_float,
    config::{Real, Status},
    simd,
};

#[inline]
pub fn train(step: &mut ModelStep<'_, '_>) -> Result<ObjectiveLoss, Status> {
    let model = &step.trainer.model;
    let dimension = model.embedding_dimension;
    let radius = context_radius_with_policy(
        step.window_rng,
        step.trainer.config.window_radius,
        step.trainer.config.context_policy,
    );
    let mut context_count = 0usize;
    step.hidden.fill(0.0);
    step.hidden_gradient.fill(0.0);

    for offset in 0..=radius * 2 {
        if let Some(position) =
            context_position(step.sentence.len(), step.sentence_position, radius, offset)
        {
            let token = step.sentence[position];
            let start = token * dimension;
            let input_row = &model.input_embeddings[start..start + dimension];
            for (coordinate, value) in input_row.iter().enumerate() {
                step.hidden[coordinate] += atomic_float::load(value);
            }
            context_count += 1;
        }
    }
    if context_count == 0 {
        return Ok(ObjectiveLoss::default());
    }
    simd::divide_in_place(step.hidden, context_count as Real);
    let loss = objective::train(
        step.trainer,
        step.target_token,
        step.learning_rate,
        step.negative_rng,
        step.hidden,
        step.hidden_gradient,
        step.observe_objective,
    )?;
    for offset in 0..=radius * 2 {
        if let Some(position) =
            context_position(step.sentence.len(), step.sentence_position, radius, offset)
        {
            let token = step.sentence[position];
            let start = token * dimension;
            let input_row = &model.input_embeddings[start..start + dimension];
            for (coordinate, value) in input_row.iter().enumerate() {
                atomic_float::add(value, step.hidden_gradient[coordinate]);
            }
        }
    }
    Ok(loss)
}
